Firestore/core/src/firebase/firestore/util/CMakeLists.txt

```diff
@@ -105,6 +105,8 @@ configure_file(
 cc_library(
   firebase_firestore_util
   SOURCES
+    async_queue.cc
```

Comment:
> [... ] The overall pattern in this folder's CMakeLists.txt is to build the resulting firebase_firestore_util library from several libraries, some of which are platform-dependent. I cannot easily do the same for the library containing AsyncQueue, because [...]

You could put them all into one library; but you'd have to add the non-apple libdispatch now-ish. Assuming you did that, you could:
1) Add the external libdisaptch as a conditional dependency, i.e.
```
if(not apple)
  list(append UTIL_DEPENDS libdispatch_dispatch_or_whatever_we_call_it)
endif()
```
2) add async_queue_libdispatch.* to SOURCES of firebase_firestore_util

Alternatively, putting this into a separate library or directory would also make sense, but doesn't mirror the obj-c code. (Which is ok, but should trigger further discussion with the larger team.) If you take this approach, rather than defining the library conditionally, I'd recommend adding the dependency conditionally. (See appendix below.)

As a temporary measure, defining the library only on apple (as you've done) is fine by me for now, though I'd recommend a TODO be added.


> Also, what async_queue library does is: [...]

One could imagine more than just the sources differing (i.e. dependencies, cflags, etc). In the extreme, this would cause the second approach to mirror the first, but with more variables. (But I would imagine that this would not be a common case.)

Looking a f/f/{auth,util}/CMakeLists.txt, we seem to be defining "private" libraries that are platform specific, and then conditionally adding them to the dependencies of "public" libraries. In other words, it would look more like this:

```
cc_library(apple_specific_thing, ...)
cc_library(nonapple_thing, ...)

if (apple)
  list(append deps apple_specific_thing)
else
  list(append deps nonapple_thing)

cc_library(
  publicish_lib
  SOURCES
    additional_platform_independent_sources.cc
  DEPENDS
    ${deps}
    additional_platform_independent_deps
)
```

Of note, the apple library would end up defined, even on non-apple platforms, though it won't actually be compiled. (This ensures the sources are present, I think, but doesn't require them to be compile-able.)

That may all be less relevant here, since I believe the source will be same on all platforms?
 
 
Appendix: Yeah, this comment is now long enough to warrant an appendix. :(

If you want to define this as a separate library (rather than being part of f_f_util), I'd recommend:

1) Create libdispatch_dispatch library (as an external dependency. Check out cmake/external/*.cmake and cmake/Find*.cmake)
2) Add the following to f/f/util/CMakeLists.txt
```
if (not apple)
  list(append async_queue_deps libdispatch_dispatch)
endif()
```
3) alter f/f/util/CMakeLists.txt and unconditionally define firebase_firestore_util_async_queue.
4) Add a ${async_queue_deps} to the DEPENDS section

Reply:
> you'd have to add the non-apple libdispatch now-ish

I don't intend to do that. I think the idea is to only use libdispatch-based implementation on Apple platforms, at least for now.

So far I have no platform-independent source files for this (the C++-based version is completely separate). Maybe it will be possible to refactor out a common part that delegates to a platform-specific "impl", but I'm not sure at this point.

Reply:
Oh, I misunderstood then; this should get easier as a result. I had assumed that this code would be used everywhere (and that we'd pull in a non-apple implementation of libdispatch such as https://github.com/nickhutchinson/libdispatch)

To be clear, this code will only ever be used on apple, right?

If so, then I think you can just
a) define the library (unconditionally)
b) conditionally add to util_depends:
```
if (apple)
  list(append util_depends f_f_u_async_queue)
endif()
```